File: cpp/3625.cpp
// Hash for pair<int,int>
struct PairHash {
    size_t operator()(const pair<int,int>& p) const noexcept {
        return (static_cast<size_t>(p.first) << 32) ^ (unsigned int)p.second;
    }
};

class Solution {
    using ll = long long;
    using Slope     = pair<int,int>; // (dy, dx) reduced
    using Intercept = pair<int,int>; // (num, den) for b = num / den
    using MidKey    = pair<int,int>; // (x1 + x2, y1 + y2)

    // Return a canonical representation of slope as (dy, dx).
    static Slope normalizeSlope(int dx, int dy) {
        if (dx == 0) // Vertical line
            return {1, 0};
        if (dy == 0) // Horizontal line
            return {0, 1};
        if (dx < 0) // Negative slope
            dx = -dx, dy = -dy;
        int g = gcd(abs(dx), abs(dy));
        dx /= g;
        dy /= g;
        return {dy, dx};
    }

    // Compute normalized intercept key for line with given slope through (x, y)
    static Intercept interceptKey(int x, int y, const Slope& s) {
        int dy = s.first;
        int dx = s.second;
        if (dx == 0) // vertical line: x = const -> use x as "intercept"
            return {x, 1};
        // y = (dy/dx)*x + b  =>  b = (y*dx - dy*x) / dx
        long long num = 1LL * y * dx - 1LL * dy * x;
        long long den = dx;
        if (den < 0) { // Negative denominator
          den = -den;
          num = -num;
        }
        long long g = gcd(abs(num), abs(den));
        if (g != 0) { // Reduce fraction
          num /= g;
          den /= g;
        }
        return {(int)num, (int)den};
    }

public:
    int countTrapezoids(vector<vector<int>>& points) {
        int n = (int)points.size();
        if (n < 4) return 0;

        // slope -> list of intercepts of segments with that slope
        unordered_map<Slope, vector<Intercept>, PairHash> slopeToIntercept;
        slopeToIntercept.reserve(1LL * n * n);

        // midpoint -> list of slopes of segments with that midpoint
        unordered_map<MidKey, vector<Slope>, PairHash> midToSlope;
        midToSlope.reserve(1LL * n * n);

        // Build all segments
        for (int i = 0; i < n; ++i) {
            int x1 = points[i][0], y1 = points[i][1];
            for (int j = i + 1; j < n; ++j) {
                int x2 = points[j][0], y2 = points[j][1];

                int dx = x2 - x1;
                int dy = y2 - y1;

                Slope k = normalizeSlope(dx, dy);
                Intercept b = interceptKey(x1, y1, k);

                slopeToIntercept[k].push_back(b);

                MidKey mid = {x1 + x2, y1 + y2};
                midToSlope[mid].push_back(k);
            }
        }

        long long ans = 0;

        // First pass: count pairs of segments with same slope but different intercept
        for (auto &entry : slopeToIntercept) {
            auto &vec = entry.second;
            if (vec.size() < 2) continue;

            unordered_map<Intercept, int, PairHash> cnt;
            cnt.reserve(vec.size());
            for (const auto &b : vec) 
                ++cnt[b];

            long long sum = 0;
            for (auto &kv : cnt) {
                long long c = kv.second;
                ans += sum * c;
                sum += c;
            }
        }

        // Second pass: subtract parallelogram overcounts via midpoints
        for (auto &entry : midToSlope) {
            auto &vec = entry.second;
            if (vec.size() < 2) continue;

            unordered_map<Slope, int, PairHash> cnt;
            cnt.reserve(vec.size());
            for (const auto &k : vec) 
              cnt[k]++;

            long long sum = 0;
            for (auto &kv : cnt) {
                long long c = kv.second;
                ans -= sum * c;
                sum += c;
            }
        }

        return (int)ans;
    }
};

File: go/3625.go
type Slope struct {
	Dy, Dx int
}

type Intercept struct {
	Num, Den int
}

type MidKey struct {
	X, Y int
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func normalizeSlope(dx, dy int) Slope {
	if dx == 0 { // Vertical line
		return Slope{1, 0}
	}
	if dy == 0 { // Horizontal line
		return Slope{0, 1}
	}
	if dx < 0 { // Negative slope
		dx, dy = -dx, -dy
	}
	g := gcd(abs(dx), abs(dy))
	dx /= g
	dy /= g
	return Slope{dy, dx}
}

func interceptKey(x, y int, slope Slope) Intercept {
	dy, dx := slope.Dy, slope.Dx
	if dx == 0 { // vertical line: x = const -> use x as "intercept"
		return Intercept{x, 1}
	}
	// y = (dy/dx)*x + b  =>  b = (y*dx - dy*x) / dx
	num := y*dx - dy*x
	den := dx
	if den < 0 { // Negative denominator
		den, num = -den, -num
	}
	g := gcd(abs(num), abs(den))
	if g != 0 { // Reduce fraction
		num /= g
		den /= g
	}
	return Intercept{num, den}
}

func countTrapezoids(points [][]int) int {
	n := len(points)
	if n < 4 {
		return 0
	}

	// slope -> list of intercepts of segments with that slope
	slopeToIntercept := make(map[Slope][]Intercept)

	// midpoint -> list of slopes of segments with that midpoint
	midToSlope := make(map[MidKey][]Slope)

	// Build all segments
	for i := 0; i < n; i++ {
		x1, y1 := points[i][0], points[i][1]
		for j := i + 1; j < n; j++ {
			x2, y2 := points[j][0], points[j][1]

			dx := x2 - x1
			dy := y2 - y1

			k := normalizeSlope(dx, dy)
			b := interceptKey(x1, y1, k)

			slopeToIntercept[k] = append(slopeToIntercept[k], b)

			mid := MidKey{x1 + x2, y1 + y2}
			midToSlope[mid] = append(midToSlope[mid], k)
		}
	}

	ans := int64(0)

	// First pass: count pairs of segments with same slope but different intercept
	for _, vec := range slopeToIntercept {
		if len(vec) < 2 {
			continue
		}

		cnt := make(map[Intercept]int)
		for _, b := range vec {
			cnt[b]++
		}

		sum := int64(0)
		for _, c := range cnt {
			ans += sum * int64(c)
			sum += int64(c)
		}
	}

	// Second pass: subtract parallelogram overcounts via midpoints
	for _, vec := range midToSlope {
		if len(vec) < 2 {
			continue
		}

		cnt := make(map[Slope]int)
		for _, k := range vec {
			cnt[k]++
		}

		sum := int64(0)
		for _, c := range cnt {
			ans -= sum * int64(c)
			sum += int64(c)
		}
	}

	return int(ans)
}

